File: sortingAlgo/mergeSort.cpp
#include<iostream>
using namespace std;

void mereg(int arr[],int l,int mid,int r){
    //size of temp array
    int n1=mid-l+1;
    int n2 = r- mid;
    // temp array
    int a[n1];
    int b[n2];

    for(int i=0;i<n1;i++){
        a[i]=arr[l+i];
    }
    for(int j=0;j<n2;j++){
        b[j]=arr[mid+1+j];
    }

    int i=0;
    int j=0;
    int k=l;
      
    while (i<n1 && j<n2)
      {
        if(a[i]<b[j]){
            arr[k]=a[i];
            k++;
            i++;
        } 
        else{
            arr[k]=b[j];
            k++; 
            j++;
        }
    }

    while (i<n1)
    {
        arr[k]=a[i];
        k++;
        i++;   
    }
    while (j<n2)
    {
        arr[k]=b[j];
        k++;
        j++;   
    }
      

}

void mergeSort(int arr[],int l,int r){
    if(l<r){
        int mid=(l+r)/2;
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        mereg(arr,l,mid,r);
    }
}

main(){
    int arr[]={5,4,3,2,1};
    mergeSort(arr,0,4);
    for(int i=0; i<5; i++){    
        cout<<arr[i]<<" ";
    }
    return 0;

}; 

File: _Interview_Basic/Algorithms/mergeSort.java
public class mergeSort {
    


    public static void mergeSorted(int arr[],int si ,int ei){
        if(si>=ei){
            return;
        }
        int mid = si + (ei-si)/2;
        mergeSorted(arr,si,mid);
        mergeSorted(arr,mid+1,ei);
        merge(arr,si,mid,ei);
    }

    public static void merge(int arr[],int si, int mid ,int ei){
        int i = si;
        int j=mid+1;
        int k=0;
        int temp[]= new int[ei-si+1];

        while (i<=mid && j<=ei) {
            if(arr[i]>arr[j]){
                temp[k]=arr[j];
                j++;
            }else{
                temp[k]=arr[i];
                i++;
            }
            k++;
        }

        while (i<=mid) {
            temp[k++]=arr[i++];
        }
        while (j<=ei) {
            temp[k++]=arr[j++];  
        }


        for(k=0,i=si;k<temp.length;k++,i++ ){
            arr[i]=temp[k];  //copying the sorted elements back to original array
        }

    }


    public static void display(int arr[]){
        for(int i=0;i<arr.length;i++){
            System.out.print(arr[i]+" ");
        }
    }

    public static void main(String[] args) {
        int arr[] = {5,7,2,1,4,3};
        mergeSorted(arr,0,arr.length-1);
        display(arr);
    }
}

